### PR TITLE
EH-Flink: Support AKS service principal provisioning by environment variable

### PR DESCRIPTION
**Pull request by kaizimmerm**
Allow to provide a service principal as alternative to auto creation see https://docs.microsoft.com/en-gb/azure/aks/kubernetes-service-principal

### CODE REVIEW

File: eventhubskafka-flink-eventhubskafka/create-solution.sh [hunk 1]
```diff
@@ -135,16 +135,19 @@ echo "Steps to be executed: $STEPS"
...
 if [ "$FLINK_PLATFORM" == "hdinsight" ]; then
-  echo ". HDInsight       => VM: $HDINSIGHT_WORKER_SIZE, Workers: $HDINSIGHT_WORKERS"
+  echo ". HDInsight           => VM: $HDINSIGHT_WORKER_SIZE, Workers: $HDINSIGHT_WORKERS"
 else
-  echo ". AKS             => VM: $AKS_VM_SIZE, Workers: $AKS_NODES"
+  echo ". AKS                 => VM: $AKS_VM_SIZE, Workers: $AKS_NODES"
+fi
+echo ". Flink               => AKS nodes: $AKS_NODES x $AKS_VM_SIZE, Parallelism: $FLINK_PARALLELISM"
+echo ". Simulators          => $SIMULATOR_INSTANCES"
+if [[ ! -z ${AD_SP_APP_ID+x} && ! -z ${AD_SP_SECRET+x} && -n "$AD_SP_APP_ID" && -n "$AD_SP_SECRET" ]]; then
```
Comment (algattik) [inline eventhubskafka-flink-eventhubskafka/create-solution.sh]: can you explain the syntax and why variables are accessed twice?
In other places we use something more compact: https://github.com/Azure-Samples/streaming-at-scale/blob/master/components/azure-dataexplorer/create-dataexplorer.sh#L16
```
if [ -n "${servicePrincipalId:-}" ]; then
```

File: eventhubskafka-flink-eventhubskafka/kubernetes/provision-aks-flink-cluster.sh [hunk 2]
```diff
@@ -17,7 +17,13 @@ echo 'creating AKS cluster'
 echo ". name: $AKS_CLUSTER"
 
 if ! az aks show --name $AKS_CLUSTER --resource-group $RESOURCE_GROUP >/dev/null 2>&1; then
-az aks create --name $AKS_CLUSTER --resource-group $RESOURCE_GROUP --node-count $AKS_NODES -s $AKS_VM_SIZE -k $AKS_KUBERNETES_VERSION --generate-ssh-keys -o tsv >> log.txt
+  # if SP vars are set and not empty
+  if [[ ! -z ${AD_SP_APP_ID+x} && ! -z ${AD_SP_SECRET+x} && -n "$AD_SP_APP_ID" && -n "$AD_SP_SECRET" ]]; then
+    echo ". service-principal: $AD_SP_APP_ID"
+    az aks create --name $AKS_CLUSTER --resource-group $RESOURCE_GROUP --node-count $AKS_NODES -s $AKS_VM_SIZE -k $AKS_KUBERNETES_VERSION --generate-ssh-keys --service-principal $AD_SP_APP_ID  --client-secret $AD_SP_SECRET -o tsv >> log.txt
+  else
+    az aks create --name $AKS_CLUSTER --resource-group $RESOURCE_GROUP --node-count $AKS_NODES -s $AKS_VM_SIZE -k $AKS_KUBERNETES_VERSION --generate-ssh-keys -o tsv >> log.txt
```
Comment (algattik) [inline eventhubskafka-flink-eventhubskafka/kubernetes/provision-aks-flink-cluster.sh]: How about incorporating the change from Julien?

```
az ad sp create-for-rbac --skip-assignation
az aks create --service-principal SPN_ID --client-secret SPN_SECRET
```

File: eventhubskafka-flink-eventhubskafka/kubernetes/provision-aks-flink-cluster.sh [hunk 3]
```diff
@@ -27,6 +33,8 @@ AKS_CLIENT_ID=$(az aks show --resource-group $RESOURCE_GROUP --name $AKS_CLUSTER
 # Get the ACR registry resource id
 ACR_ID=$(az acr show --name $ACR_NAME --resource-group $RESOURCE_GROUP --query "id" --output tsv)
 
+echo 'assign role ro SP for ACR pull'
```
Comment (algattik) [inline eventhubskafka-flink-eventhubskafka/kubernetes/provision-aks-flink-cluster.sh]: typo

File: eventhubskafka-flink-eventhubskafka/create-solution.sh (hunk 1 above)
Comment (kaizimmerm) [reply eventhubskafka-flink-eventhubskafka/create-solution.sh]: Your are right, looks much more elegant.

File: eventhubskafka-flink-eventhubskafka/kubernetes/provision-aks-flink-cluster.sh (hunk 2 above)
Comment (kaizimmerm) [reply eventhubskafka-flink-eventhubskafka/kubernetes/provision-aks-flink-cluster.sh]: This PR change makes only sense for scenarios where the call `az ad sp create-for-rbac --skip-assignation` has been done upfront (e.g. by a different person/department) as the person that runs the sample might not have permission to create a SP.

The output of the SP create command is what will go into `AD_SP_APP_ID` and `AD_SP_SECRET`.

**Pull request merged**